packages/services/api/src/modules/organization/providers/resource-selector.ts

```diff
@@ -0,0 +1,213 @@
+import { Storage } from '../../shared/providers/storage';
+
+/**
+ * Responsible for auth checks.
+ * Talks to Storage.
+ */
+@Injectable({
+  scope: Scope.Operation,
+  global: true,
+})
+export class ResourceSelector {
+  constructor(
+    private storage: Storage,
+    private session: Session,
+  ) {}
+
+  private async _assertResourceSelectorAdminPermissions(organizationId: string) {
+    const canDos = await Promise.all([
+      this.session.canPerformAction({
+        organizationId,
+        action: 'member:modify',
+        params: {
+          organizationId,
+        },
+      }),
+      this.session.canPerformAction({
+        organizationId,
+        action: 'oidc:modify',
+        params: {
+          organizationId,
+        },
+      }),
+      this.session.canPerformAction({
+        organizationId,
+        action: 'accessToken:modify',
+        params: {
+          organizationId,
+        },
+      }),
+    ]);
+
+    if (!canDos.some(canDo => canDo)) {
+      throw new AccessError('Insufficient permissions.');
+    }
+  }
+
+  async getProjectsFromOrganizationForResourceSelector(
+    organization: Organization,
+    intent: GraphQLSchema.ResourceSelectorIntentType,
+  ) {
+    let projects = await this.storage.getProjects({ organizationId: organization.id });
+
+    if (intent === 'ADMIN') {
+      await this._assertResourceSelectorAdminPermissions(organization.id);
+    } else {
+      const filteredProjects: typeof projects = [];
+      for (const project of projects) {
+        if (
+          false ===
+          (await this.session.canPerformAction({
+            action: 'project:describe',
+            organizationId: project.orgId,
+            params: {
+              organizationId: project.orgId,
+              projectId: project.id,
+            },
+          }))
+        ) {
+          continue;
+        }
+        filteredProjects.push(project);
+      }
+      projects = filteredProjects;
+    }
+
+    return projects.map(project => ({
+      projectId: project.id,
+      organizationId: project.orgId,
+      slug: project.slug,
+      type: project.type,
+    }));
+  }
+
+  async getProjectFromOrganizationForResourceSelector(
+    organization: Organization,
+    projectId: string,
+    intent: GraphQLSchema.ResourceSelectorIntentType,
+  ) {
+    const project = await this.storage.getProjectById(projectId);
+
+    if (!project) {
+      return null;
+    }
+
+    if (intent === 'ADMIN') {
+      await this._assertResourceSelectorAdminPermissions(organization.id);
+    } else {
+      // otherwise: default check if user has access on this resource
+      await this.session.assertPerformAction({
+        action: 'project:describe',
+        organizationId: organization.id,
+        params: {
+          organizationId: organization.id,
+          projectId: project.id,
+        },
+      });
+    }
+
+    if (!project || project.orgId !== organization.id) {
+      return null;
```

Comment:
This seems like a strange case. I wonder if we should throw here or at least log?